### PR TITLE
Add neural tube defect and live birth observer

### PR DESCRIPTION
**Pull request by KjellSwedin**

### CODE REVIEW

File: src/vivarium_conic_lsff/model_specifications/model_spec.in [hunk 1]
```diff
@@ -8,7 +8,7 @@ components:
             - SIS('diarrheal_diseases')
             - SIR_fixed_duration('measles', '10')
             - SIS('lower_respiratory_infections')
-            - NeonatalSWC_without_incidence('neural_tube_defects')
```
Comment (collijk) [inline src/vivarium_conic_lsff/model_specifications/model_spec.in]: You've taken neural tube defects out of the model?

File: src/vivarium_conic_lsff/components/observers.py [hunk 2]
```diff
@@ -132,4 +132,98 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
                                                         year=event_time.year)
     base_filter = QueryString('')
     transition_count = get_group_counts(transitioned_pop, base_filter, base_key, config, age_bins)
-    return transition_count
\ No newline at end of file
+    return transition_count
+
+
+
+class LiveBirthWithNTDObserver:
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py:140]: For specialized observers, I'll typically drop the abstraction layer in configuration defaults.  So have 
```
configuration_defaults = {
    'metrics: {
        'live_births_with_ntds': {
            'by_year': True,
            'by_sex': True,
        }
    }
}
```

And then skip the whole __init__ method.

File: src/vivarium_conic_lsff/components/observers.py [hunk 3]
```diff
@@ -132,4 +132,98 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+    @property
+    def name(self):
+        return f'live_birth_with_ntd_observer'
+
+    def setup(self, builder):
+        self.config = builder.configuration['metrics'][f'{self.disease}_observer'].to_dict()
+        self.live_birth_counts = Counter()
+        self.with_disease = Counter()
+        tmp = builder.configuration['time']['start']
+        self._reference_time = pd.Timestamp(tmp['year'], tmp['month'], tmp['day'])
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: `reference_time` is ambiguous.  Try `self._simulation_start = pd.Timestamp(**builder.configuration.time.start.to_dict())`

File: src/vivarium_conic_lsff/components/observers.py [hunk 4]
```diff
@@ -132,4 +132,98 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+
+    def __repr__(self):
+        return f"DiseaseObserver({self.disease})"
+
+
+def get_live_births(pop, config, reference_time, year):
+    retval = {}
+    filter = QueryString(f'alive == "alive" and entrance_time >= "{reference_time}"')
+    base_key = get_output_template(False, **config).substitute(
+        measure=f'live_births', year=year if config['by_year'] else None)
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Always pass the year.  The output template is smart and will ignore it if the config is not by year.

File: src/vivarium_conic_lsff/components/observers.py [hunk 5]
```diff
@@ -132,4 +132,98 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+    if config['by_sex']:
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Use https://github.com/ihmeuw/vivarium_public_health/blob/master/src/vivarium_public_health/metrics/utilities.py#L267 for this.  Add {'by_age': False} to your config and pass None or an empty df to get_group_counts for the age bins.

File: src/vivarium_conic_lsff/components/observers.py [hunk 6]
```diff
@@ -132,4 +132,98 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+
+
+def get_with_ntd_births(pop, config, reference_time, year):
+    retval = {}
+    filter = QueryString(f'alive == "alive"'
+            f' and {project_globals.NTD_MODEL_NAME} == "{project_globals.NTD_MODEL_NAME}"'
+            f' and entrance_time >= "{reference_time}"')
+    base_key = get_output_template(False, **config).substitute(
+        measure=f'born_with_ntd', year=year if config['by_year'] else None)
+    if config['by_sex']:
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Same as above.  This duplicates what get_group_counts does.

**Review by collijk — CHANGES_REQUESTED** [review]
See comments.

File: src/vivarium_conic_lsff/components/observers.py (hunk 6 above)
Comment (KjellSwedin) [reply src/vivarium_conic_lsff/components/observers.py]: OK, I looked at get_group_counts() and it wasn't clear that I could use it without the age bins. I'll switch things to use it.

File: src/vivarium_conic_lsff/components/observers.py [hunk 7]
```diff
@@ -132,4 +132,72 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+            'live_births_with_ntds': {
+                'by_year': True,
+                'by_sex': True,
+            }
+        }
+    }
+
+    @property
+    def name(self):
+        return f'live_birth_with_ntd_observer'
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Use the project globals here too.

File: src/vivarium_conic_lsff/components/observers.py [hunk 8]
```diff
@@ -132,4 +132,72 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+
+    def setup(self, builder):
+        self.disease = project_globals.NTD_MODEL_NAME
+        self.config = builder.configuration['metrics']['live_births_with_ntds'].to_dict()
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: I'd update your config here directly to include `by_age: False`.

File: src/vivarium_conic_lsff/components/observers.py [hunk 9]
```diff
@@ -132,4 +132,72 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        self.live_birth_counts = Counter()
+        self.with_disease = Counter()
+        tmp = builder.configuration['time']['start']
+        self._step_start_time = pd.Timestamp(tmp['year'], tmp['month'], tmp['day'])
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Still think unpacking builder.configuration.time.start here directly.  It's why the keys are structured the way they are and allows us the flexibility to specify any args a timestamp will take without code changes.

File: src/vivarium_conic_lsff/components/observers.py [hunk 10]
```diff
@@ -132,4 +132,72 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+            columns_required.append('sex')
+        self.population_view = builder.population.get_view(columns_required)
+
+        builder.value.register_value_modifier('metrics', self.metrics)
+        builder.event.register_listener('collect_metrics', self.on_collect_metrics)
+
+    def on_collect_metrics(self, event):
+        pop = self.population_view.get(event.index)
+        self.live_birth_counts.update(
+            get_live_births(pop, self.config, self._step_start_time, event.time.year))
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Do these in two lines:
```
live_births_this_step = get_live_births(...)
self.live_birth_counts.update(live_births_this_step).
```
Two reasons:
1. It's easier to use a debugger if get_live_births breaks or (the worse case) you are unintentionally overwriting keys.
2. Our pattern for stratification is to get the counts this step, update the dictionary keys with our stratification criteria, then update the observer count dict, so we'll have to change it anyway.

File: src/vivarium_conic_lsff/components/observers.py [hunk 11]
```diff
@@ -132,4 +132,72 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        metrics.update(self.live_birth_counts)
+        metrics.update(self.with_disease)
+        return metrics
+
+    def __repr__(self):
+        return f"DiseaseObserver({self.disease})"
+
+
+def get_live_births(pop, config, reference_time, year):
+    filter = QueryString(f'alive == "alive" and entrance_time >= "{reference_time}"')
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: you'll want `entrance_time == current_sim_time` or you'll count the same births over and over.

File: src/vivarium_conic_lsff/components/observers.py [hunk 12]
```diff
@@ -132,4 +132,73 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        return project_globals.NTD_OBSERVER
+
+    def setup(self, builder):
+        self.disease = project_globals.NTD_MODEL_NAME
+        self.config = builder.configuration['metrics'][project_globals.NTD_OBSERVER].to_dict()
+        self.config['by_age'] = False
+        self.live_birth_counts = Counter()
+        self.with_disease = Counter()
+        tmp = builder.configuration['time']['start']
+        self._step_start_time = pd.Timestamp(tmp['year'], tmp['month'], tmp['day'])
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Instead, grab  `self.clock = builder.time.clock()`

File: src/vivarium_conic_lsff/components/observers.py [hunk 13]
```diff
@@ -132,4 +132,73 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        builder.value.register_value_modifier('metrics', self.metrics)
+        builder.event.register_listener('collect_metrics', self.on_collect_metrics)
+
+    def on_collect_metrics(self, event):
+        pop = self.population_view.get(event.index)
+        live_births_this_step = get_live_births(pop, self.config, self._step_start_time, event.time.year)
+        self.live_birth_counts.update(live_births_this_step)
+        births_with_ntds_this_step = get_with_ntd_births(pop, self.config, self._step_start_time, event.time.year)
+        self.with_disease.update(births_with_ntds_this_step)
+        self._step_start_time = pd.Timestamp(event.time.year, event.time.month, event.time.day)
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Delete this line

File: src/vivarium_conic_lsff/components/observers.py [hunk 14]
```diff
@@ -132,4 +133,95 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+    @property
+    def name(self):
+        return project_globals.NTD_OBSERVER
+
+    def setup(self, builder):
+        self.disease = project_globals.NTD_MODEL_NAME
+        self.config = builder.configuration['metrics'][project_globals.NTD_OBSERVER].to_dict()
+        self.config['by_age'] = False
+
+        tmp = builder.configuration['time']['start']
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py:162]:
```
self._sim_start = pd.Timestamp(**builder.configuration.time.start.to_dict())
self._sim_end = pd.Timestamp(**builder.configuration.time.end.to_dict())
```

File: src/vivarium_conic_lsff/components/observers.py [hunk 15]
```diff
@@ -132,4 +133,95 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        self._sim_start = pd.Timestamp(tmp.year, tmp.month, tmp.day)
+        tmp = builder.configuration['time']['end']
+        self._sim_end = pd.Timestamp(tmp.year, tmp.month, tmp.day)
+
+        columns_required = ['alive', 'dead', f'{self.disease}', 'entrance_time', 'tracked', 'untracked']
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: shouldn't have an untracked column.  I'm surprised this doesn't fail.  No untracked column exists.

File: src/vivarium_conic_lsff/components/observers.py [hunk 16]
```diff
@@ -132,4 +133,95 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+    births
+        All births and births with neural tube defects present.
+    """
+    base_filter = QueryString('')
+    base_key = get_output_template(**config)
+    time_spans = get_time_iterable(config, sim_start, sim_end)
+
+    births = {}
+    for year, (t_start, t_end) in time_spans:
+        start = max(sim_start, t_start)
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py:212]: Seems annoying this isn't handled by `get_time_iterable`

File: src/vivarium_conic_lsff/components/observers.py [hunk 17]
```diff
@@ -132,4 +133,95 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+        end = min(sim_end, t_end)
+        born_in_span = pop.query(f'entrance_time >= "{start}" and entrance_time < "{end}"')
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py]: Readability thing:  I think it's a good idea to pick exactly one ordering operator and always use it. I typically only use `<`.  To me `'{start} <= entrance_time and entrance_time < {end}' ` is much clearer.

File: src/vivarium_conic_lsff/components/observers.py [hunk 18]
```diff
@@ -132,4 +133,95 @@ def get_transition_count(pop, config, disease, transition, event_time, age_bins)
...
+
+        cat_year_key = base_key.substitute(measure='live_births', year=year)
+        filter = base_filter
+        group_births = get_group_counts(born_in_span, filter, cat_year_key, config, pd.DataFrame())
```
Comment (collijk) [inline src/vivarium_conic_lsff/components/observers.py:218]: make an `empty_age_bins` variable and assign the empty dataframe to it for clarity.

**Review by collijk — APPROVED** [review]
Nitpick comments.  Address what you'd like.  Nicely done.

**Pull request merged**